core/trino-main/src/main/java/io/trino/sql/planner/LocalDynamicFilterConsumer.java

```diff
@@ -13,105 +13,115 @@
 import static com.google.common.base.Preconditions.checkState;
 import static com.google.common.base.Verify.verify;
 import static com.google.common.collect.ImmutableMap.toImmutableMap;
+import static io.trino.spi.predicate.TupleDomain.columnWiseUnion;
 import static java.util.Objects.requireNonNull;
 import static java.util.function.Function.identity;
 
 public class LocalDynamicFilterConsumer
         implements DynamicFilterSourceConsumer
 {
-    private static final int PARTITION_COUNT_INITIAL_VALUE = -1;
     // Mapping from dynamic filter ID to its build channel indices.
     private final Map<DynamicFilterId, Integer> buildChannels;
-
     // Mapping from dynamic filter ID to its build channel type.
     private final Map<DynamicFilterId, Type> filterBuildTypes;
-
     private final List<Consumer<Map<DynamicFilterId, Domain>>> collectors;
+    private final long domainSizeLimitInBytes;
 
     // Number of build-side partitions to be collected, must be provided by setPartitionCount
     @GuardedBy("this")
-    private int expectedPartitionCount = PARTITION_COUNT_INITIAL_VALUE;
-
+    private Integer expectedPartitionCount;
     @GuardedBy("this")
-    private boolean collected;
-
-    // The resulting predicates from each build-side partition.
-    @Nullable
+    private int collectedPartitionCount;
     @GuardedBy("this")
-    private List<TupleDomain<DynamicFilterId>> partitions;
+    private volatile boolean collected;
 
-    public LocalDynamicFilterConsumer(
-            Map<DynamicFilterId, Integer> buildChannels,
-            Map<DynamicFilterId, Type> filterBuildTypes,
-            List<Consumer<Map<DynamicFilterId, Domain>>> collectors)
+    private final Queue<TupleDomain<DynamicFilterId>> summaryDomains = new ConcurrentLinkedQueue<>();
+
+    public LocalDynamicFilterConsumer(Map<DynamicFilterId, Integer> buildChannels, Map<DynamicFilterId, Type> filterBuildTypes, List<Consumer<Map<DynamicFilterId, Domain>>> collectors, DataSize domainSizeLimit)
     {
         this.buildChannels = requireNonNull(buildChannels, "buildChannels is null");
         this.filterBuildTypes = requireNonNull(filterBuildTypes, "filterBuildTypes is null");
         verify(buildChannels.keySet().equals(filterBuildTypes.keySet()), "filterBuildTypes and buildChannels must have same keys");
-
-        requireNonNull(collectors, "collectors is null");
-        checkArgument(!collectors.isEmpty(), "collectors is empty");
-        this.collectors = collectors;
-        this.partitions = new ArrayList<>();
+        this.collectors = ImmutableList.copyOf(requireNonNull(collectors, "collectors is null"));
+        this.domainSizeLimitInBytes = requireNonNull(domainSizeLimit, "domainSizeLimit is null").toBytes();
     }
 
     @Override
-    public void addPartition(TupleDomain<DynamicFilterId> tupleDomain)
+    public void addPartition(TupleDomain<DynamicFilterId> domain)
     {
+        if (collected) {
```

Comment:
It looks confusing to have `collected` declared as `GuardedBy("this")` and then also be a volatile so that you can use it without guarding by "this".
Can we move `summaryDomains.add(domain)` just before `collectedPartitionCount++` and run `compactSummaryDomains` after the critical section ?

Reply:
It has to opportunistically try to compact domains before critical section to keep the algorithm correct. I think in other places we also have other volatile variables declared as `GuardedBy("this")` while doing opportunistic checks outside the lock.